### PR TITLE
Multipass typings

### LINKED ISSUE #96: Foreach-binding with object syntax fails
**Current behavior**
When using `foreach: { data: x, as: 'x' }`, the resulting binding context type will be unkown (unless 'as' is const type).

**Expected behavior**
A new binding context with a property 'x' defined.

**Steps to reproduce**
```html
<span data-bind="foreach: { data: x, as: 'x' }"></span>
```

**Suggested solution**
Since we have a similar functionality in the 'let' binding-handler, we should reuse the same typing support for the 'foreach' case.

Transform all return expressions where an object is returned (instead of the normal array type) to a type similar to what the 'let'-binding uses.

Example: the binding `{ data: x, as: 'x' }` should be transformed to `{ 'x': x }` internally.
The 'as' property value should become the new property name to replace the 'data' property.

### PR DESCRIPTION
**Pull request by maskmaster**
A heavily reworked version of the code generation, utilizing more of the typescript functionality to do the scaffolding (the initial template file), type injections in multiple passes, tracking source-map changes through mulitple generations of generated files.

Review is preferably done on a commit-by-commit basis. Sorry for the big diff.

Main features:
* Sourcemaps working end-to-end
* Multi-pass typing-injection to make sure that dependent types are correctly inferred
* Fixes the foreach-bindings with object parameters ({ data: x. as: y})

Fixes #96

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+         "type": "pwa-node",
+         "name": "Run Command Line",
+         "program": "${workspaceFolder}/build/cli/cli.js",
+         "args": [".\\Areas\\portal\\Portlets\\Returns\\views\\simpleReturn.html"],
+         "cwd": "c:\\Seco\\mypages\\app\\portal\\portal.web",
+         "request": "launch",
+         "skipFiles": [
+            "<node_internals>/**"
+         ]
+      },
```
Comment (tscpp) [inline .vscode/launch.json:14]: ????

File: demo/resources/view.html [hunk 2]
```diff
@@ -1,6 +1,23 @@
...
-<div data-bind="koko: [prop]">
+<!-- ko-viewmodel: import Vymodell from './viewmodel' -->
+<!-- ko-bindinghandler: import koko1 from './bindinghandler' -->
+<!-- ko-bindinghandler: import * as koko2 from './bindinghandler' -->
+<!-- ko-bindinghandler: import typeof koko3 from './bindinghandler' -->
+<!-- ko-bindinghandler: import typeof * as koko4 from './bindinghandler' -->
+<!-- ko-bindinghandler: import { koko5 } from './bindinghandler' -->
+<!-- ko-bindinghandler: import { koko as KOKO6 } from './bindinghandler' -->
+<!-- ko-bindinghandler: import { typeof koko7 } from './bindinghandler' -->
+<!-- ko-bindinghandler: import { typeof koko as KOKO8 } from './bindinghandler' -->
```
Comment (tscpp) [inline demo/resources/view.html:9]: Too many binding handler imports.

File: demo/resources/view.html [hunk 3]
```diff
@@ -1,6 +1,23 @@
-<!-- ko-viewmodel: import default from './viewmodel' -->
-<!-- ko-bindinghandler: import koko from './bindinghandler' -->
-
-<div data-bind="koko: [prop]">
+<!-- ko-viewmodel: import Vymodell from './viewmodel' -->
```
Comment (tscpp) [inline demo/resources/view.html:1]:
```suggestion
<!-- ko-viewmodel: import ViewModel from './viewmodel' -->
```

File: demo/src/lint.test.ts [hunk 4]
```diff
@@ -1,39 +1,39 @@
...
 	console.log('')
 }
 
 async function main() {
 	const filename = path.join(__dirname, '../resources/view.html')
 	const relFilename = path.relative(process.cwd(), filename).replace(/\\/g, '/')
-
 	const textDoc = fs.readFileSync(filename).toString()
+	if (!textDoc)
+		throw new Error(`Cannot open file '${filename}'`)
```
Comment (tscpp) [inline demo/src/lint.test.ts:23]: `fs` throws exceptions, not empty strings. Use try statement.

File: demo/src/lint.test.ts [hunk 5]
```diff
@@ -1,39 +1,39 @@
...
-		const typescriptEmit = await program.compile(filename, document)
-
-		const diagnostics = program.getDiagnostics().concat(typescriptEmit.getDiagnostics())
+		const diagnostics = program.getDiagnostics()
 
 		if (!fs.existsSync(path.join(__dirname, '../temp')))
 			fs.mkdirSync(path.join(__dirname, '../temp'))
 
-		fs.writeFileSync(path.join(__dirname, '../temp/output.ts'), typescriptEmit.rawSource)
+		// fs.writeFileSync(path.join(__dirname, '../temp/output.ts'), typescriptEmit.rawSource)
```
Comment (tscpp) [inline demo/src/lint.test.ts:36]: Remove if no further use.

File: lib/resources/context.d.ts [hunk 6]
```diff
@@ -12,23 +12,13 @@ type LastElement<Arr extends [...any]> = Arr extends [...infer U, infer T] ? T :
...
-	[]
-
 /** Used to signal type argument error */
 type BindingContextRequired = 'Argument must be a Binding Context'
 
-/** Create an S-Expression from a BindinContext-hierarchy.
- * Example: ChildBindingContext<A, ChildBindingContext<B, RootBindingContext<C>>> -> [A, [B, [C, []]]] */
-type CreateSExpr<BC> =
-	BC extends ChildBindingContext<infer V, infer U> ? [V, CreateSExpr<U>] :
+export type XX<BC> =
```
Comment (tscpp) [inline lib/resources/context.d.ts:18]: XX doesn't really mean much.  Consider adding TSDoc or improve name.

File: spec/src/spec.ts [hunk 7]
```diff
@@ -1,3 +1,4 @@
+import path = require('path')
```
Comment (tscpp) [inline spec/src/spec.ts:1]: Why is require import necessary here?

File: src/parser/bindingDOM.ts [hunk 8]
```diff
@@ -67,7 +67,7 @@ export class BindingExpression {
 }
 
 export class Document {
-	public constructor(public bindings: Binding[], public viewmodelReferences: ViewModelNode[], public bindingHandlerReferences: BindingHandlerImportNode[]) { }
+	public constructor(public rootBinding: Binding, public viewmodelReferences: ViewModelNode[], public bindingHandlerReferences: BindingHandlerImportNode[]) { }
```
Comment (tscpp) [inline src/parser/bindingDOM.ts:70]: Is there a root binding? If not, could this be improved?

**Review by tscpp — CHANGES_REQUESTED** [review]

**Comment by tscpp** [comment]
Add changes to CHANGELOG.md

File: demo/resources/view.html [hunk 9]
```diff
@@ -1,6 +1,23 @@
...
+</div>
+<div data-bind="foreach: [prop]">
+    <span data-bind="text: $root.propp"></span>
+    <div data-bind="foreach: { data: [$data], as: 'kropp' }">
+        <div data-bind="koko1: [kropp]">
+        </div>
+    </div>
+</div>
+<div data-bind="koko1: [prop]">
 </div>
```
Comment (tscpp) [inline demo/resources/view.html:23]:
```suggestion
<!-- ko-viewmodel: import ViewModel from './viewmodel' -->
<!-- ko-bindinghandler: import custom from './bindinghandler' -->

<div data-bind="foreach: { data: [prop], as: 'body' }">
    <div data-bind="text: body"></div>
</div>
<div data-bind="foreach: [prop]">
    <span data-bind="text: $root.prop"></span>
    <div data-bind="foreach: { data: [$data], as: 'body' }">
        <div data-bind="custom: [body]"></div>
    </div>
</div>
<div data-bind="custom: [prop]"></div>
```

File: src/parser/bindingDOM.ts (hunk 8 above)
Comment (maskmaster) [reply src/parser/bindingDOM.ts:70]: Yes. All documents have a root
Excerpt from the `createDocument` function.
```typescript
const root = new Binding(new BindingName('root', undefined as unknown as Location), new BindingExpression('', undefined as unknown as Location))
```

File: demo/resources/view.html (hunk 2 above)
Comment (maskmaster) [reply demo/resources/view.html:9]: This is to try out/document all import types.

Comment (tscpp) [reply demo/resources/view.html:9]: It should be tested in spec. I'll allow it because of #110.

File: spec/src/spec.ts (hunk 7 above)
Comment (maskmaster) [reply spec/src/spec.ts]: By mistake. VSCode  auto generated that one.

File: demo/src/lint.test.ts (hunk 4 above)
Comment (maskmaster) [reply demo/src/lint.test.ts:23]: Removed

**Comment by tscpp** [comment]
Sorry for the conflicts...

**Review by tscpp — APPROVED** [review]

**Pull request merged**